src/OrbitGl/QtTextRenderer.cpp

```diff
@@ -124,14 +117,27 @@ void QtTextRenderer::AddText(const char* text, float x, float y, float z, TextFo
       idx_min = candidate_idx;
     }
   }
-  const size_t fitting_chars_count = idx_min;
-  stored_text_.emplace_back(text_as_string.substr(0, fitting_chars_count).c_str(),
-                            transformed.xy[0], transformed.xy[1], transformed.z, formatting);
+  text_as_string = text_as_string.substr(0, idx_min);
+  float y_offset = GetYOffsetFromAlignment(
+      formatting.valign, GetStringHeight(text_as_string.c_str(), formatting.font_size));
+  const float single_line_height = GetStringHeight(".", formatting.font_size);
+  const std::vector<std::string> lines = absl::StrSplit(text_as_string, '\n');
+  for (const auto& line : lines) {
+    const float width = GetStringWidth(line.c_str(), formatting.font_size);
+    const float x_offset = GetXOffsetFromAlignment(formatting.halign, width);
+    stored_text_[transformed.z].emplace_back(
+        line.c_str(), std::lround(transformed.xy[0] + x_offset),
+        std::lround(transformed.xy[1] + y_offset), std::lround(width),
+        std::lround(single_line_height), formatting);
+    y_offset += single_line_height;
+  }
+
   if (out_text_pos != nullptr) {
-    *out_text_pos = transformed.xy;
-    (*out_text_pos)[0] +=
+    (*out_text_pos)[0] =
```

Comment:
We are calculating the Width and the Height again in this call for both pos and size. Don't you like to calculate this as well in the previous iteration?

Reply:
move GetStringWidth, GetStringHeight for entire text to the top of the method.